Get flash_attn to compile for CUDA 11.6 linux nightly build

cpuhrsch opened the pull request:
This PR only attempts to get this code to compile for all archs so that we can dispatch to it in https://github.com/pytorch/pytorch/pull/84653

drisspg commented on aten/src/ATen/native/transformers/transformer.cpp:
```diff
@@ -107,20 +108,29 @@ Tensor transformer_encoder_layer_forward(
        embed_dim / num_heads == 64 ||
        embed_dim / num_heads == 128)) {
-     TORCH_WARN_ONCE("USING FLASH ATTENTION WITH NT");
+     TORCH_WARN_ONCE("transformer_encoder_layer_forward is using flash attention.");
```
maybe add flash_attention is in prototype or something

drisspg commented on aten/src/ATen/CMakeLists.txt:
```diff
@@ -156,11 +156,10 @@ file(GLOB native_transformers_hip_cpp "native/transformers/hip/*.cpp")
-file(GLOB flash_attention_cuda_cpp "native/transformers/cuda/flash_attn/*.cpp")
-
-if(USE_FLASH_ATTENTION)
+if(USE_CUDA AND NOT USE_ROCM)
```
Why are we no longer using the USE_FLASH_ATTENTION flag to gate whether or not we build the sources?

cpuhrsch replied:
To enable the code at runtime it needs to have been compiled. It'll be compiled for NVIDIA CUDA only.

drisspg replied:
But I thought we can't build flash unless we are on an A100? So CI is going to complain

drisspg replied:
Got it, I thought for some reason we weren't able to compile this code when not on both modern CUDA and newer cuda arch. I wonder if nvcc flags specifying arch will not work for CI/CD

cpuhrsch replied:
Yeah we might need to update those flags so that CI builds for 8.0 as well. Let's wait for build failures again before we do that.

malfet replied:
Perhaps we can still use this flag, but just set it to TRUE by default in a top-level CMakeLists.txt? (When option is defined)

drisspg commented on aten/src/ATen/native/transformers/cuda/flash_attn/fmha_api.h:
```diff
@@ -6,6 +6,7 @@
 
 namespace fmha {
 
+TORCH_API
```
Why do we expose this symbol now? Was this a problem with the version before?

cpuhrsch replied:
There were build errors that show linking issues for the cpp tests. Adding this should make the symbols visible.

drisspg reviewed (commented):
The flag makes sense just not sure if we can by default build flash with cuda ?

cpuhrsch commented:
@drisspg - as long as the compiler is modern enough it should be possible. We'll see if we need to guard against certain compiler versions.

drisspg reviewed (approved):
Assuming CI/CD is good I think this is good. Might be good to get @malfet's opinion

cpuhrsch commented:
I'm worried about the CUDA context size due to this change. cc @ngimel

ngimel commented:
You can check the context size?

cpuhrsch commented:
Yep, on it. Sorry, I only wanted to add you for visibility @ngimel. Depending on the measurements here we might need to follow the linalg approach of creating a separate library and load it at runtime only when needed. This matters especially as we're adding more and more kernels.

cpuhrsch commented:
I see an increase of around 6MiB or 0.74% due to this change.

```
$ nvidia-smi --query-gpu=utilization.memory,memory.total,memory.free,memory.used --format=csv -i 0
utilization.memory [%], memory.total [MiB], memory.free [MiB], memory.used [MiB]
0 %, 40960 MiB, 39540 MiB, 813 MiB
```

vs.

```
$ nvidia-smi --query-gpu=utilization.memory,memory.total,memory.free,memory.used --format=csv -i 0
utilization.memory [%], memory.total [MiB], memory.free [MiB], memory.used [MiB]
0 %, 40960 MiB, 39534 MiB, 819 MiB
```

Running

```
import torch
a = torch.randn(3).cuda()
```

malfet commented on aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:
```diff
@@ -13,9 +13,7 @@
 // TODO Consider moving all flash_attention code, nested tensor included to
 // Transformer library
 
-#ifdef USE_FLASH_ATTENTION
 #include <ATen/native/transformers/cuda/flash_attn/fmha_api.h>
-#endif
```
Again, I propose we keep it as is (otherwise it'll be a hell to fix internal builds)

malfet reviewed (changes requested):
I propose we keep `USE_FLASH_ATTENTION` define for now, but flip it from FALSE to TRUE from top-level CMakeLists.txt (and use `cmake_dependent_option to make it false if USE_CUDA is set to false) https://github.com/pytorch/pytorch/blob/94b67f4cd8dc1ab5f7add5f006f7f3fd988b8ecf/CMakeLists.txt#L728

malfet commented:
6Mb increase sounds fine (though undesirable, but we can think about it later and 11.7 should help there as well)
But I though this code would not build for older CUDA architectures (i.e. needs sm_70+), but if it works as is, than we are fine

malfet commented on CMakeLists.txt:
```diff
@@ -725,7 +725,10 @@ set(BUILD_ONEDNN_GRAPH OFF)
 
 # Moved this cmake set option down here because CMAKE_CUDA_COMPILER_VERSION is not avaialble until now
-option(USE_FLASH_ATTENTION "Whether to build the flash_attention kernel for scaled dot product attention" OFF)
+option(USE_FLASH_ATTENTION "Whether to build the flash_attention kernel for scaled dot product attention" ON)
+if(MSVC OR (CMAKE_CUDA_COMPILER_VERSION VERSION_LESS 11.6))
+  set(USE_FLASH_ATTENTION OFF)
+ENDIF()
```
Use `cmake_dependen_option`
```suggestion
cmake_dependent_option(USE_FLASH_ATTENTION "Whether to build the flash_attention kernel for scaled dot product attention" ON "NOT MSVC AND NOT CMAKE_CUDA_COMPILER_VERSION VERSION_LESS 11.6" OFF)
```

malfet commented on aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:
```diff
@@ -13,9 +13,7 @@
 // TODO Consider moving all flash_attention code, nested tensor included to
 // Transformer library
 
-#ifdef USE_FLASH_ATTENTION
```
You want to keep it in place, aren't you?

malfet commented on CMakeLists.txt:
```diff
@@ -725,12 +725,16 @@ set(BUILD_ONEDNN_GRAPH OFF)
 ENDIF()
 
 
+
```
Nit
```suggestion
```

cpuhrsch committed:
Get flash_attn to compile for CUDA 11.6 linux nightly build (#84941)

Summary:
This PR only attempts to get this code to compile for all archs so that we can dispatch to it in https://github.com/pytorch/pytorch/pull/84653

Pull Request resolved: https://github.com/pytorch/pytorch/pull/84941

Reviewed By: malfet

Differential Revision: D39632988

Pulled By: cpuhrsch

fbshipit-source-id: f5752b5a0fd1c24756b4eaafdcf4511afbf6c50d
